### PR TITLE
ENH: Mitutoyo excel spread sheet reader

### PR DESCRIPTION
**Pull request by jotelha**
Reads file at once. Extracts acquisition date and a few standard roughness metrics. No tests, no documentation yet.

Would produce some hypothetical output
```
List of channels:
[<SurfaceTopography.IO.Reader.ChannelInfo object at 0x7f6a62fcbaf0>]

Channel name:
default

Physical sizes:
(0.48,)

Number of grid points:
(960,)

Dimensions:
1

Info:
{'channel_name': 'default', 'roughness_metrics': 0     {'key': 'Ra', 'value': 0.078, 'unit': 'µm'}
1     {'key': 'Rq', 'value': 0.096, 'unit': 'µm'}
2     {'key': 'Rz', 'value': 0.392, 'unit': 'µm'}
3    {'key': 'Rpk', 'value': 0.155, 'unit': 'µm'}
4    {'key': 'Rvk', 'value': 0.084, 'unit': 'µm'}
Name: 0, dtype: object, 'acquisition_time': '2020-07-26 00:00:00', 'unit': 'um'}

Topography physical sizes:
(0.48,)

Topography rms height:
0.10166847079311534
```

when

```
from SurfaceTopography import open_topography

# get a handle to the file ("reader")
reader = open_topography("mitutoyo.xlsx", format='mitutoyo')   # you can find this file in the folder 'tests/file_format_examples'

# each file has a list of channels (one or more)
print("List of channels:")
print(reader.channels)  # returns list of channels
ch = reader.channels[0]  # first channel, alternatively use ..
ch = reader.default_channel  # .. - one of the channels is the "default" channel

# each channel has some defined meta data
print("")
print("Channel name:")
print(ch.name)  # channel name
print("")
print("Physical sizes:")
print(ch.physical_sizes)  # lateral dimensions
print("")
print("Number of grid points:")
print(ch.nb_grid_pts)  # number of grid points
print("")
print("Dimensions:")
print(ch.dim)  # number of dimensions (1 or 2)
print("")
print("Info:")
print(ch.info)  # more metadata, e.g. 'unit' if unit was given in file

# you can get a topography from a channel
topo = ch.topography()   # here meta data from the file is taken

# each topography has a rich set of methods and properties for meta data and analysis
print("")
print("Topography physical sizes:")
print(topo.physical_sizes)  # lateral dimension
print("")
print("Topography rms height:")
print(topo.rms_height_from_area())  # Root mean square of heights
```

run  on some arbitrary Mitutoyo excel spread sheet file

### CODE REVIEW

File: SurfaceTopography/IO/Mitutoyo.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    '''
+
+    def __init__(self, fobj):
+        """
+        Open Excel spread sheet produced by Mitutoyo SurfTest surface roughness
+        testers.
+
+        The reader expects a line scan by positions and heights in columns
+        5 and 6 (E, F) and tries to extract standard roughness metrics from
+        column 1 (A) on the sheet 'DATA'. The reader extracts exüects the
```
Comment (pastewka) [inline SurfaceTopography/IO/Mitutoyo.py:58]: Typo "exüects"

**Review by pastewka — CHANGES_REQUESTED** [review]
Please remove the "WIP" qualifier. Please provide tests.

**Comment by jotelha** [comment]
If no WIP qualifier, how to mark a pull request that's there for discussion and not finalized yet?

A few notes:

- Stitched together from bits and pieces of other readers and tests. Do we have some specs documentation somewhere on how a reader should look like?
- Reads whole excel file in constructor
- Any exception in constructor reraised as IOError
- Always assumes um as unit, does not try to determine unit from file (the format does not offer a unique location to look up the units), should likely assume mm for positions and um for heights per default
- For now, assumes uniform line scan from zero to maximum position, although the excel format would hypothetically allow for nonuniform positions.
- Constructs such info object:

```json
{'acquisition_time': '2022-10-10 00:00:00',
 'channel_name': 'default',
 'roughness_metrics': [{'key': 'Ra', 'unit': 'µm', 'value': 0.018},
                       {'key': 'Rq', 'unit': 'µm', 'value': 0.027},
                       {'key': 'Rz', 'unit': 'µm', 'value': 0.191},
                       {'key': 'Rpk', 'unit': 'µm', 'value': 0.052},
                       {'key': 'Rvk', 'unit': 'µm', 'value': 0.489}],
 'unit': 'um'}

```
- Test uses minimal mock xlsx file with randomly generated values

**Comment by pastewka** [comment]
That said, let's just assume for the units are fixed and correct as implemented.

**Comment by jotelha** [comment]
> > If no WIP qualifier, how to mark a pull request that's there for discussion and not finalized yet?
> 
> If not yet finalized, keep WIP. I thought it was finalized since you requested a review.
> 
Inserted again
> > A few notes:
> > 
> > * Stitched together from bits and pieces of other readers and tests. Do we have some specs documentation somewhere on how a reader should look like?
> 
> It is not formally documented, but feel free to add that documentation to the code. (I think it should go into `Reader.py`.)
> 
Put a few open questions as comments in the mitutoyo reader code at every point where I just copied bits without really understanding whether these are needed or what they do. Answering those would lead to such documentation :)

> > * Reads whole excel file in constructor
> 
> That is fine. I think there is no way to avoid this. Text parser do the same thing.
> 
> > * Any exception in constructor reraised as IOError
> 
> Okay. What type of exception may occur? We should make sure only actual I/O error are raised as IOErrors. Otherwise debugging becomes a mess.

I don't know what kind of exceptions pandas read_excel might throw and could not find any documentation on that, but now catch exceptions from these pandas calls only.

> 
> > * Always assumes um as unit, does not try to determine unit from file (the format does not offer a unique location to look up the units), should likely assume mm for positions and um for heights per default
> 
> We should write to Mitutoyo to ask whether this is the always the case.
> 
Tries to infer units from the little information found within the excel file now.
> > * For now, assumes uniform line scan from zero to maximum position, although the excel format would hypothetically allow for nonuniform positions.
> 
> Please add a check that raises an exception if the position are nonuniform. You can simple check that `np.diff` on the x-position yields a constant.

Added that check and construct either uniform or nonuniform profile, but either I am doing something wrong in the construction of `UniformLinescan` vs `NonuniformLinescan` or there is something wrong in the `rms_height_from_profile` of one of them, see tests https://github.com/ContactEngineering/SurfaceTopography/blob/6ac93ead766f78b6b95a3337094d8c40d7993000/test/IO/test_mitutoyo.py#L37 vs https://github.com/ContactEngineering/SurfaceTopography/blob/6ac93ead766f78b6b95a3337094d8c40d7993000/test/IO/test_mitutoyo.py#L50
The mock xlsx input files are almost the same with a little modification to the position in row 100.



> 
> > * Constructs such info object:
> > 
> > ```json
> > {'acquisition_time': '2022-10-10 00:00:00',
> >  'channel_name': 'default',
> >  'roughness_metrics': [{'key': 'Ra', 'unit': 'µm', 'value': 0.018},
> >                        {'key': 'Rq', 'unit': 'µm', 'value': 0.027},
> >                        {'key': 'Rz', 'unit': 'µm', 'value': 0.191},
> >                        {'key': 'Rpk', 'unit': 'µm', 'value': 0.052},
> >                        {'key': 'Rvk', 'unit': 'µm', 'value': 0.489}],
> >  'unit': 'um'}
> > ```
> 
> That looks good to me, except that I think channel name and unit do not need to be in that dictionary.
That's again just copied from some other reader, i.e. https://github.com/ContactEngineering/SurfaceTopography/blob/6ac93ead766f78b6b95a3337094d8c40d7993000/SurfaceTopography/IO/Text.py#L246-L250
> 
> > * Test uses minimal mock xlsx file with randomly generated values
> 
> Sounds good.

**Comment by pastewka** [comment]
> > > * Stitched together from bits and pieces of other readers and tests. Do we have some specs documentation somewhere on how a reader should look like?
> > 
> > 
> > It is not formally documented, but feel free to add that documentation to the code. (I think it should go into `Reader.py`.)
> 
> Put a few open questions as comments in the mitutoyo reader code at every point where I just copied bits without really understanding whether these are needed or what they do. Answering those would lead to such documentation :)

I've updated the reader and added more documentation to the base class. Please check if that doc is clearer now.

> > > * Any exception in constructor reraised as IOError
> > 
> > 
> > Okay. What type of exception may occur? We should make sure only actual I/O error are raised as IOErrors. Otherwise debugging becomes a mess.
> 
> I don't know what kind of exceptions pandas read_excel might throw and could not find any documentation on that, but now catch exceptions from these pandas calls only.

The way it is handled now is good. Only reraise during actual read, not during parsing of the metadata.

> Added that check and construct either uniform or nonuniform profile, but either I am doing something wrong in the construction of `UniformLinescan` vs `NonuniformLinescan` or there is something wrong in the `rms_height_from_profile` of one of them, see tests

The test works on my computer.

> > That looks good to me, except that I think channel name and unit do not need to be in that dictionary.
> > That's again just copied from some other reader, i.e. https://github.com/ContactEngineering/SurfaceTopography/blob/6ac93ead766f78b6b95a3337094d8c40d7993000/SurfaceTopography/IO/Text.py#L246-L250

The channel name is not part of the info dictionary. In those other readers, the name is actually removed before constructing the topography object.

**Review by pastewka — DISMISSED** [review]

**Comment by pastewka** [comment]
Please have a look at it @jotelha - happy to have this merged

@mcrot - this introduces a pandas dependency

**Comment by jotelha** [comment]
> > Added that check and construct either uniform or nonuniform profile, but either I am doing something wrong in the construction of `UniformLinescan` vs `NonuniformLinescan` or there is something wrong in the `rms_height_from_profile` of one of them, see tests
> 
> The test works on my computer.
> 
The tests work, but only because they test against different values of `rms_height_from_profile`, 1.69 for `UniformLineScan`,
https://github.com/ContactEngineering/SurfaceTopography/blob/6ac93ead766f78b6b95a3337094d8c40d7993000/test/IO/test_mitutoyo.py#L37 and 1.37 for `NonuniformLineScan`, https://github.com/ContactEngineering/SurfaceTopography/blob/6ac93ead766f78b6b95a3337094d8c40d7993000/test/IO/test_mitutoyo.py#L50

The data files are essentially the same with a slightly off-grid x position in row 100, 0.05 as uniform vs 0.0501 as nonuniform.

Looking at the implementations of https://github.com/ContactEngineering/SurfaceTopography/blob/34816d954c4886440ba10eadda4a4b8b721526f3/SurfaceTopography/Uniform/ScalarParameters.py#L39-L61 vs
https://github.com/ContactEngineering/SurfaceTopography/blob/34816d954c4886440ba10eadda4a4b8b721526f3/SurfaceTopography/Nonuniform/ScalarParameters.py#L37-L72 that discrepancy must somehow originate in the weighted contribution of segments approximated piece-wise linearly, is that possible?

**Comment by pastewka** [comment]
How many grid points does the example file have?

There are a couple of differences between uniform and nonuniform:
* For uniform topography, the code essentially assumes that the height is constant over the pixel width. If you have _nx_ points, then the overall length of the scan is _nx dx_
* Nonuniform topographies are interpreted as points connect by lines. The rms calculation actually makes use of this linear interpolation. The over length of the scan is then _(nx-1) dx_
* If your file has small _nx_, you will get noticably different values

**Comment by jotelha** [comment]
I see, it's 960 grid points, and the data series does not start at zero x, but at the first non-zero x value 0.005 and ends at 0.48 mm, i.e. height values are assigned to the "end" of each "pixel" in the format. I use random noise between ~ -0.24 um and 0.35 um as height test data, the source of the discrepancy will be here in the arbitrary jumps between one h_i and the other h_i+1.

**Comment by pastewka** [comment]
Somehow the test stalls. This is annoying.

**Comment by pastewka** [comment]
Sorry, what is still missing is adding the Mitutoyo reader to the binary file formats in `test_io.py`

**Comment by jotelha** [comment]
Now, there are
```
FAILED test/IO/test_io.py::test_reader_topography_same[/mnt/dat2/git/ContactEngineering/SurfaceTopography/test/file_format_examples/mitutoyo_nonuniform_mock.xlsx] - assert (480.0,) == (479.5,)
FAILED test/IO/test_io.py::test_periodic_flag[/mnt/dat2/git/ContactEngineering/SurfaceTopography/test/file_format_examples/mitutoyo_nonuniform_mock.xlsx] - AssertionError: /mnt/dat2/git/ContactEngineering/Sur...
FAILED test/IO/test_io.py::test_to_netcdf[/mnt/dat2/git/ContactEngineering/SurfaceTopography/test/file_format_examples/mitutoyo_mock.xlsx] - assert <SurfaceTopog...x7f3bb1232490> == <SurfaceTopog...x7f3bb3bda...
FAILED test/IO/test_io.py::test_to_netcdf[/mnt/dat2/git/ContactEngineering/SurfaceTopography/test/file_format_examples/mitutoyo_nonuniform_mock.xlsx] - assert <SurfaceTopog...x7f3bb7f9d430> == <SurfaceTopog.....
```
left failing, and my guess is these all have to do with the physical_size vs grid in the mitutoyo format,

```
____________________________________ test_reader_topography_same[/mnt/dat2/git/ContactEngineering/SurfaceTopography/test/file_format_examples/mitutoyo_nonuniform_mock.xlsx] _____________________________________

fn = '/mnt/dat2/git/ContactEngineering/SurfaceTopography/test/file_format_examples/mitutoyo_nonuniform_mock.xlsx'

    @pytest.mark.parametrize('fn', text_example_file_list + text_example_without_size_file_list + binary_example_file_list +
                             binary_without_stream_support_example_file_list)
    def test_reader_topography_same(fn):
        """
        Tests that properties like physical sizes, units and nb_grid_pts are
        the same in the ChannelInfo and the loaded topography.
        """
    
        reader = open_topography(fn)
    
        for channel in reader.channels:
            foo_str = reader.format() + "-%d" % (channel.index,)  # unique for each channel
            topography = channel.topography(
                physical_sizes=(1, 1) if channel.physical_sizes is None
                else None,
                info=dict(foo=foo_str))
            assert channel.nb_grid_pts == topography.nb_grid_pts
    
            # some checks on info dict in channel and topography
            assert topography.info['foo'] == foo_str
            if channel.unit is not None or topography.unit is not None:
                assert channel.unit == topography.unit
                assert channel.info['unit'] == topography.unit
                assert channel.unit == topography.info['unit']
    
            if channel.physical_sizes is not None:
>               assert channel.physical_sizes == topography.physical_sizes
E               assert (480.0,) == (479.5,)
E                 At index 0 diff: 480.0 != 479.5
E                 Use -v to get more diff
```

**Comment by jotelha** [comment]
Why do I get those test errors locally, but they run through here?

**Comment by jotelha** [comment]
With the last commit, all tests run locally as well, but that means a little modiication to the NonuniformLineScan class resolving https://github.com/ContactEngineering/SurfaceTopography/issues/230, do you want that?

**Comment by pastewka** [comment]
I've pushed my changes.

**Comment by pastewka** [comment]
Hi @jotelha - I saw that you added the periodic flag to `NonuniformLineScan`. Please remove it again. This requires further thinking.

**Review by pastewka — APPROVED** [review]

**Pull request merged**